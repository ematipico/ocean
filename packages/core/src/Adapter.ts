export abstract class Adapter {
	abstract build(): void;

	abstract loadConfiguration(): void;
}
