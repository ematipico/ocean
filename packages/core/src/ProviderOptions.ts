export interface EnvironmentVariables {
	[x: string]: string;
}

export interface ProviderOptions {
	runtimeVersion?: '10' | '12';
	gatewayKey?: string;
	memorySize?: string;
	timeout?: string;
	environmentVariables?: EnvironmentVariables;
}
