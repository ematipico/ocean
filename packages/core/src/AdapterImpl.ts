import { Adapter } from './Adapter';

export class AdapterImpl implements Adapter {
	build(): void {
		throw new Error('Implement me');
	}

	loadConfiguration(): void {
		throw new Error('Implement me');
	}
}
