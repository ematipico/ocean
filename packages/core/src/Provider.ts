import { Mapping } from './Adapter';

export abstract class Provider {
	abstract setMappings(mappings: Mapping[]): void;

	abstract run(): void;

	abstract async asyncRun(): Promise<void>;
}
