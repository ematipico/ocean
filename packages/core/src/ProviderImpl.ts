import { Provider } from './Provider';
import { Mapping } from './Adapter';
import { ProviderOptions } from './ProviderOptions';

export interface ProviderConstructor {
	mappings: Mapping[];
	pathToFiles: string;
	providerOptions: ProviderOptions;
}

export class ProviderImpl<ProviderConfiguration> implements Provider {
	providerConfiguration: ProviderConfiguration;
	protected pathToFiles: string;
	protected mappings: Mapping[];

	constructor({ mappings, pathToFiles }: ProviderConstructor) {
		this.mappings = mappings;
		this.pathToFiles = pathToFiles;
	}

	setMappings(mappings: Mapping[]): void {
		this.mappings = mappings;
	}

	async asyncRun(): Promise<void> {
		return Promise.reject('Implement me');
	}

	run(): void {
		throw new Error('Implement me');
	}
}
