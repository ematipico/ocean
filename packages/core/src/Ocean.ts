import {MissingOceanConfiguration} from "./errors/MissingOceanConfiguration";
import {Adapter} from "./Adapter";
import {createNewAdapter} from "./factory";

export enum AdapterEnum {
	NuxtJS = '@ocean/nuxt-adapter',
	NextJS = '@ocean/next-adapter',
}
export enum ProviderEnum {
	AWS = '@ocean/aws-provider',
}

export interface OceanConfiguration {
	adapter: AdapterEnum;
	provider: ProviderEnum;
}

export class Ocean<AdapterInterface extends Adapter = Adapter> {
	adapter: AdapterInterface;
	provider: ProviderEnum;

	constructor({ adapter, provider }: OceanConfiguration) {
		if (!adapter || !provider) {
			throw new MissingOceanConfiguration();
		}

		this.adapter = createNewAdapter<AdapterInterface>(adapter);
		this.provider = provider;
	}

	async run() {
		await Promise.resolve();
	}
}
