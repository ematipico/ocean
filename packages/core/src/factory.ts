import { AdapterEnum, ProviderEnum } from './Ocean';
import { NoAdapterFound } from './errors/NoAdapterFound';
import { NoProviderFound } from './errors/NoProviderFound';

export function createNewAdapter<AdapterInterface>(adapter: AdapterEnum): AdapterInterface {
	try {
		const { Adapter } = require(adapter);
		return new Adapter();
	} catch (e) {
		throw new NoAdapterFound(adapter, e);
	}
}

export function createNewProvider(provider: ProviderEnum) {
	try {
		const { Provider } = require(provider);
		return new Provider();
	} catch (e) {
		throw new NoProviderFound(provider, e);
	}
}
