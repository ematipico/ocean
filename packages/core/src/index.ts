import { Ocean } from './Ocean';
import { AdapterImpl } from './AdapterImpl';

export { Ocean, AdapterImpl };
