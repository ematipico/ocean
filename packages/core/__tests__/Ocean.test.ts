import {Ocean} from "../src/Ocean";

describe('@ocean/core', () => {
    it('should throw an error for missing configuration', () => {
        const a = () => {
            // @ts-ignore
            new Ocean();
        }

        expect(a).toThrowError();
        expect(a).toThrowErrorMatchingSnapshot();
    });

    it('should throw an error for empty configuration', () => {
        const a = () => {
            // @ts-ignore
            new Ocean({});
        }

        expect(a).toThrowError();
        expect(a).toThrowErrorMatchingSnapshot();
    });

    it('should throw an error for adapter not found', () => {
        const a = () => {
            // @ts-ignore
            new Ocean({
                // @ts-ignore
                adapter: 'fake-package',
                // @ts-ignore
                provider: 'fake-provider'
            });
        }

        expect(a).toThrowError();
        expect(a).toThrowErrorMatchingSnapshot();
    });

    it('should throw an error for provider found', () => {
        const a = () => {
            // @ts-ignore
            new Ocean({
                // @ts-ignore
                adapter: '@ocean/nextjs-adapter',
                // @ts-ignore
                provider: 'fake-provider'
            });
        }

        expect(a).toThrowError();
        expect(a).toThrowErrorMatchingSnapshot();
    });

    it('should not throw an error when correct packages are passed', () => {
        const a = () => {
            // @ts-ignore
            new Ocean({
                // @ts-ignore
                adapter: '@ocean/nextjs-adapter',
                // @ts-ignore
                provider: '@ocean/aws-provider'
            });
        }

        expect(a).not.toThrowError();
    });
})
