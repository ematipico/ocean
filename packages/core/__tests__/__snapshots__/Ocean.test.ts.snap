// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@ocean/core should not throw an error when correct packages are passed 1`] = `"Cannot find the package \\"@ocean/nextjs-adapter\\" or the package doesn't export the \\"Adapter\\" class"`;

exports[`@ocean/core should throw an error for adapter not found 1`] = `"Cannot find the package \\"fake-package\\" or the package doesn't export the \\"Adapter\\" class"`;

exports[`@ocean/core should throw an error for empty configuration 1`] = `"Adapter or Provider are missing"`;

exports[`@ocean/core should throw an error for missing configuration 1`] = `"Cannot destructure property \`adapter\` of 'undefined' or 'null'."`;

exports[`@ocean/core should throw an error for provider found 1`] = `"Cannot find the package \\"@ocean/nextjs-adapter\\" or the package doesn't export the \\"Adapter\\" class"`;
